File: emr-frontend-v2/src/constants/systemRoles.ts
export const SYSTEM_ROLES = {
  admin: 'ADMIN',
  engineer: 'ENGINEER',
  technician: 'TECHNICIAN',
  departmentUser: 'DEPARTMENT_USER',
  accountant: 'ACCOUNTANT',
  procurement: 'PROCUREMENT'
} as const

export const DEVICE_ROLES = [
  SYSTEM_ROLES.admin,
  SYSTEM_ROLES.engineer,
  SYSTEM_ROLES.technician,
  SYSTEM_ROLES.departmentUser,
  SYSTEM_ROLES.accountant,
  SYSTEM_ROLES.procurement
]

export const REPAIR_LIST_ROLES = DEVICE_ROLES

export const REPAIR_CREATE_ROLES = DEVICE_ROLES

export const REPAIR_EXECUTE_ROLES = [
  SYSTEM_ROLES.admin,
  SYSTEM_ROLES.engineer,
  SYSTEM_ROLES.technician
]

export const MAINTENANCE_ROLES = [
  SYSTEM_ROLES.admin,
  SYSTEM_ROLES.engineer,
  SYSTEM_ROLES.technician,
  SYSTEM_ROLES.procurement
]

export const MAINTENANCE_EXECUTE_ROLES = [
  SYSTEM_ROLES.admin,
  SYSTEM_ROLES.technician
]

export const USER_MANAGEMENT_ROLES = [SYSTEM_ROLES.admin]

export const TRANSFER_ROLES = DEVICE_ROLES

export const ASSET_OUTPUT_ROLES = DEVICE_ROLES

export const PROCUREMENT_ROLES = [
  SYSTEM_ROLES.admin,
  SYSTEM_ROLES.engineer,
  SYSTEM_ROLES.technician,
  SYSTEM_ROLES.departmentUser,
  SYSTEM_ROLES.accountant,
  SYSTEM_ROLES.procurement
]

export function normalizeRole(role?: string | null): string | null {
  if (!role) return null

  const value = role.trim().toLowerCase().replace(/[\s-]+/g, '_')
  if (value === 'admin') return SYSTEM_ROLES.admin
  if (value === 'engineer' || value === 'ky_thuat') return SYSTEM_ROLES.engineer
  if (value === 'technician' || value === 'vat_tu_thiet_bi') return SYSTEM_ROLES.technician
  if (value === 'department_user' || value === 'deptuser' || value === 'departmentuser' || value === 'khoa_phong') return SYSTEM_ROLES.departmentUser
  if (value === 'accountant' || value === 'ke_toan' || value === 'ketoan') return SYSTEM_ROLES.accountant
  if (value === 'procurement' || value === 'mua_hang' || value === 'muahang') return SYSTEM_ROLES.procurement
  return role.trim().toUpperCase()
}

File: EMR-backend/src/Domain/Security/SystemRoles.cs
using System.Globalization;
using System.Text;

namespace EMR.Domain.Security
{
    public static class SystemRoles
    {
        public const string Admin = "ADMIN";
        public const string Engineer = "ENGINEER";
        public const string Technician = "TECHNICIAN";
        public const string DepartmentUser = "DEPARTMENT_USER";
        public const string Accountant = "ACCOUNTANT";
        public const string Procurement = "PROCUREMENT";

        public const string AdminName = "Admin";
        public const string EngineerName = "Engineer";
        public const string TechnicianName = "Technician";
        public const string DepartmentUserName = "Department User";
        public const string AccountantName = "Ke toan";
        public const string ProcurementName = "Mua hang";

        public static readonly IReadOnlyList<string> All = new[]
        {
            Admin,
            Engineer,
            Technician,
            DepartmentUser,
            Accountant,
            Procurement
        };

        public static readonly IReadOnlyDictionary<string, string> DisplayNames = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase)
        {
            [Admin] = AdminName,
            [Engineer] = EngineerName,
            [Technician] = TechnicianName,
            [DepartmentUser] = DepartmentUserName,
            [Accountant] = AccountantName,
            [Procurement] = ProcurementName
        };

        private static readonly IReadOnlyDictionary<string, string> AliasMap = new Dictionary<string, string>(StringComparer.Ordinal)
        {
            [NormalizeKey(Admin)] = Admin,
            [NormalizeKey(AdminName)] = Admin,
            [NormalizeKey(Engineer)] = Engineer,
            [NormalizeKey(EngineerName)] = Engineer,
            [NormalizeKey("Ky thuat")] = Engineer,
            [NormalizeKey(Technician)] = Technician,
            [NormalizeKey(TechnicianName)] = Technician,
            [NormalizeKey("Vat tu thiet bi")] = Technician,
            [NormalizeKey(DepartmentUser)] = DepartmentUser,
            [NormalizeKey(DepartmentUserName)] = DepartmentUser,
            [NormalizeKey("DeptUser")] = DepartmentUser,
            [NormalizeKey("DepartmentUser")] = DepartmentUser,
            [NormalizeKey("Khoa phong")] = DepartmentUser,
            [NormalizeKey(Accountant)] = Accountant,
            [NormalizeKey(AccountantName)] = Accountant,
            [NormalizeKey("KeToan")] = Accountant,
            [NormalizeKey(Procurement)] = Procurement,
            [NormalizeKey(ProcurementName)] = Procurement,
            [NormalizeKey("MuaHang")] = Procurement
        };

        public static string Normalize(string? role)
        {
            return ToCode(role);
        }

        public static string ToCode(string? role)
        {
            return TryNormalize(role, out var normalized)
                ? normalized
                : role?.Trim().ToUpperInvariant() ?? string.Empty;
        }

        public static bool TryNormalize(string? role, out string normalized)
        {
            normalized = string.Empty;
            if (string.IsNullOrWhiteSpace(role))
            {
                return false;
            }

            return AliasMap.TryGetValue(NormalizeKey(role), out normalized!);
        }

        public static bool IsCanonical(string? role)
        {
            if (string.IsNullOrWhiteSpace(role))
            {
                return false;
            }

            return All.Contains(ToCode(role), StringComparer.OrdinalIgnoreCase);
        }

        public static string GetDisplayName(string? role)
        {
            var code = ToCode(role);
            return DisplayNames.TryGetValue(code, out var displayName)
                ? displayName
                : role?.Trim() ?? string.Empty;
        }

        public static string ToPermissionKey(string? role)
        {
            var normalized = ToCode(role);
            return string.IsNullOrWhiteSpace(normalized)
                ? string.Empty
                : NormalizeKey(normalized).Replace(" ", ".", StringComparison.Ordinal);
        }

        private static string NormalizeKey(string value)
        {
            var normalized = RemoveDiacritics(value.Trim()).ToLowerInvariant();
            var builder = new StringBuilder(normalized.Length);
            var previousWasSpace = false;

            foreach (var ch in normalized)
            {
                if (char.IsLetterOrDigit(ch))
                {
                    builder.Append(ch);
                    previousWasSpace = false;
                    continue;
                }

                if (!previousWasSpace)
                {
                    builder.Append(' ');
                    previousWasSpace = true;
                }
            }

            return builder.ToString().Trim();
        }

        private static string RemoveDiacritics(string value)
        {
            var normalized = value.Normalize(NormalizationForm.FormD);
            var builder = new StringBuilder(normalized.Length);

            foreach (var ch in normalized)
            {
                var category = CharUnicodeInfo.GetUnicodeCategory(ch);
                if (category != UnicodeCategory.NonSpacingMark)
                {
                    builder.Append(ch);
                }
            }

            return builder.ToString().Normalize(NormalizationForm.FormC);
        }
    }
}
